### DISCUSSION TITLE
Why doesn't Microcks serialize Kafka messages?

### DISCUSSION THREAD

**Discussion by CROprogrammer** [microcks/microcks#867] [Q&A] [answered]
Hi,

I'm using Microcks to send messages to Kafka. Messages are using AsyncAPI schemas. I'm also trying to use Apicurio schema registry with it (no luck currently). 
I've achived that Kafka messages are being written to the topic. I'm using `UserSignedUpAPI-asyncapi` AsyncAPI schema.

I've noticed that messages that arrive to Kafka are not serialized.
For example:
![image](https://github.com/microcks/microcks/assets/73464772/6643e839-0e80-4d52-abc9-6edbda495e12)

I want them to arrive serialized, for example:
![image](https://github.com/microcks/microcks/assets/73464772/bc2577bb-9665-4c72-8953-01fad77f7b12)

Is that possbile using Microcks?

**Comment by CROprogrammer** [comment]
I've noticed that Microcks does serialize message but not the way I was thinking. It uses Jackson ObjectMapper serialization, not Apicurio Serde of Confluent SerDes. 
Will there in the future be support for Apicurio/Confluent serde?

**Reply by lbroudoux** [reply, marked as answer]
Hey!

The binary serialization of a message depends on the message's content-type. The `UserSignedUpAPI-asyncapi` sample (see [here](https://github.com/microcks/microcks/blob/master/samples/UserSignedUpAPI-asyncapi.yml)) defines `application/json` messages so the messages are just JSON strings. However, the `UserSignedUpAPI-Avro-Ref` sample (see [here](https://github.com/microcks/microcks/blob/master/webapp/src/test/resources/io/github/microcks/util/asyncapi/user-signedup-avro-ref-asyncapi.yaml)) defines `avro/binary` messages so the messages sent will actually be serialized in binary Avro.

Microcks has 4 different types of serializations, depending on content type but also on your encoding/registry settings as explained in [this guide explaining schema registry settings](https://microcks.io/documentation/guides/avro-messaging/#1-setup-schema-registry).

1. If content type is `application/json`, the message content will be a simple JSON string (probably produced with Jackson ObjectMapper)
2. If content type is binary (`avro/binary`, `application/octet-stream` or `application/avro`), we will then look at the `defaultAvroEncoding` configuration property.
  2.1. If its value is `RAW`, the examples provided as YAML or JSON will be encoded in binary Avro by generic Avro `DatumReader` and `DatumWriter` and published directly on topic,
  2.2. If its value is `REGISTRY`, will then look at the `schemaRegistry. confluent` configuration flag. 
    1. If it's `true`, we use the `Confluent SerDes` library to write the message as well as ensure the Avro schema is registered in the connected registry
    2. If it's `false`, we use the `Apicurio Registry SerDes` library to write the message as well as ensure the Avro schema is registered in the connected registry

Hope it helps make things clearer!
